lib/child_process.js

```diff
@@ -276,33 +275,26 @@ ObjectDefineProperty(exec, promisify.custom, {
  * @returns {ChildProcess}
  */
 function execFile(file, args = [], options, callback) {
-  if (args == null) {
+  if (args != null && typeof args === 'object' && !ArrayIsArray(args)) {
+    callback = options;
+    options = args;
     args = [];
-  } else if (typeof args === 'object') {
-    if (!ArrayIsArray(args)) {
-      callback = options;
-      options = args;
-      args = [];
-    }
   } else if (typeof args === 'function') {
     callback = args;
-    options = {};
+    options = null;
     args = [];
-  } else {
-    throw new ERR_INVALID_ARG_VALUE('args', args);
   }
 
-  if (options == null) {
-    options = {};
-  } else if (typeof options === 'function') {
+  if (typeof options === 'function') {
     callback = options;
-    options = {};
-  } else if (typeof options !== 'object') {
-    throw new ERR_INVALID_ARG_VALUE('options', options);
+    options = null;
+  } else if (options != null) {
+    // TODO: Should this have allowArray: true set?
+    validateObject(options, 'options');
```

Comment:
Our tests don't seem to check what happens if it's an array, but I wonder if we should do it for consistency. We had to make this allowance in `fork()` to accommodate our own benchmark test runner. I'm inclined to confirm that an array is permitted in the current release of Node.js, then add tests, and apply them here, which will then require adding `allowArray` here.

Reply:
Confirmed that an array is tolerated currently. `node -e "child_process.execFile('ls', [], [])"` does not throw an error, but on this branch it does because we don't have `allowArray` set here.

Reply:
Hmmm...but the tests for this sort of thing in `test/parallel/test-child-process-spawn-typeerror.js` talk about confirming that `execFile()` and `fork()` have the same behavior as `spawn()`. But they don't because `spawn()` does *not* tolerate an array as an options object.

So is the correct thing to do here to preserve the current behavior (where `execFile()` and `fork()` are more lax than `spawn()`), make `spawn()` more lax, or make `execFile()` and `fork()` more strict?

Reply:
@nodejs/child_process Thoughts on this? `fork()` and `execFile()` allow an array as an options object, but `spawn()` does not. Should they be consistent? If so, which is right?

```console
$ node -e "child_process.execFile('ls', [], [])"
$ node -e "child_process.fork('index.js', [], [])"
$ node -e "child_process.spawn('ls', [], [])"
node:internal/validators:164
      throw new ERR_INVALID_ARG_TYPE(name, 'Object', value);
      ^

TypeError [ERR_INVALID_ARG_TYPE]: The "options" argument must be of type object. Received an instance of Array
```

If it matters, our current benchmark runner depends on this feature/quirk in `fork()`.

Reply:
🤔 Passing an array for the `options` parameter is not documented -- https://nodejs.org/dist/latest-v16.x/docs/api/child_process.html#child_processforkmodulepath-args-options. Therefore, I would argue that it is wrong to accept an array.

Reply:
> I would argue that it is wrong to accept an array.

That was initially my take too. The only reason I continued to accommodated it in `fork()` in this PR (so far) is because our own `benchmark/run.js` relies on this behavior. Immediately after posting my comment, though, I thought the answer was obvious:

* Update `run.js` so it doesn't use this quirk.
* Remove the accommodation from this PR, adding tests for the change.
* Document it. Make sure it gets highlighted in the changelog for 18.x. Run CITGM to see if any of the modules tested there rely on th quirk. Open PRs for any modules found to use/rely on the quirk.

Reply:
I concur.